### PR TITLE
[WIP] Active tracer cookbooks

### PR DESCRIPTION
**Pull request by naliboff**
A preliminary pull request (nowhere near ready to merge) for adding cookbooks illustrating how to use active tracers.  Two initial ideas (prm files included) are taking the stokes-sinker and composition-active cookbooks and modifying them to use active tracers in replace of active compositional fields.  

Note that this branch is based off of @gassmoeller tracer_interpolation_to_field branch, which is not yet merged.  Once those features are merged, I will rebase with respect to the master.  FYI, I will need to return to this in August as I will be out of email communication a good chunk of July.  If someone else would like to talk this sooner, feel free to delete/close the pull request.

### CODE REVIEW

**Comment by tjhei** [comment]
Make sure that you set your username and email address on the machine you commit, otherwise they won't be connected to your github account.

File: cookbooks/sinker-with-active-particles/run.sh [hunk 1]
```diff
@@ -0,0 +1,7 @@
+# aspect executable 
+ASPECT="/Users/jbnaliboff/Software/aspect/aspect/./aspect"
```
Comment (tjhei) [inline cookbooks/sinker-with-active-particles/run.sh]: Please don't include a hardcoded path like this. Take a look at the other benchmarks.

**Comment by bangerth** [comment]
You'll need to run the indentation script and also rebase to current master.

File: doc/manual/cookbooks/sinker-with-active-particles/sinker-with-active-particles.prm [hunk 2]
```diff
@@ -0,0 +1,108 @@
+set Dimension                              = 2
```
Comment (bangerth) [inline doc/manual/cookbooks/sinker-with-active-particles/sinker-with-active-particles.prm]: In the cookbook, we typically don't include the *whole* .prm file, unless essentially everything is new. Rather, we only include those parts that really are specific to what is being demonstrated. Can you reduce what you include here to the important parts? I suspect that it is based on a .prm file that has previously been discussed in the manual anyway, and then you can say that it is "based on the parameter file discussed in Section~\ref{...}, with only the following sections changed in relevant ways (the entire input file can be found in \url{...}):"

File: doc/manual/cookbooks/tracer-active/tracer-active.prm [hunk 3]
```diff
@@ -0,0 +1,133 @@
+#########################################################
+# This is a variation of the composition-passive.prm file.
+# Here, we choose a higher Rayleigh number and make the
+# density depend on the compositional variable as well.
```
Comment (bangerth) [inline doc/manual/cookbooks/tracer-active/tracer-active.prm]: same here

File: include/aspect/introspection.h [hunk 4]
```diff
@@ -294,6 +294,28 @@ namespace aspect
...
+       * compositional fields. This type is used to determine how
+       * to advect a compositional field.
+       */
+      struct FieldMethod
+      {
+          enum kind
+          {
+            continuous_fem_field = 0,
+            discontinuous_fem_field = 1,
+            particles = 2
```
Comment (bangerth) [inline include/aspect/introspection.h]: does this come from a separate pull request that should be reviewed separately?

**Comment by bangerth** [comment]
I think the reference to @gassmoeller 's branch is to #1007.

**Comment by gassmoeller** [comment]
Now that #1007 is merged you should be able to rebase this PR to the current master. You will probably have to throw away all of my commits (I have changed a few things in #1007 and rebased it before it was merged, and this will lead to conflicts for you). You can do that by updating your local master doing a `git rebase -i master` from this branch, and dropping all of my commits (make sure you keep your commit). Since this process might destroy things you can create a 'backup' of this branch before doing the rebase by creating a new branch from this one (check out this branch and just create a new one by `git checkout -b backup`). You can also cancel the rebase at any time (unless it is successful) by doing `git rebase --abort`. After all is done you can check out your rebased branch and update this pull request by doing `git push -f origin active_tracer_cookbooks` (assuming your remote link is named origin, which is the default).

**Comment by naliboff** [comment]
@gassmoeller - A messy rebase indeed.  Part way through and not clear that I'm creating simply additional conflicts.  At this point it is just easier to start a new pull request branched from the current master (as long as this ok with you).

**Comment by naliboff** [comment]
@gassmoeller - This is ready for review.  If you could take a quick look to make sure I have not said anything truly erroneous, that would be fantastic.  One other thing to consider is whether we should stick with the name "active tracer" or switch to something else.  I think "active tracer" or "active particle" can work, but some people in the community have very strong feelings about this.

File: doc/manual/manual.tex [hunk 5]
```diff
@@ -4745,6 +4745,26 @@ \subsubsection{Using tracer particles}
...
+
+This cookbook (\url{cookbooks/composition-active-tracers.prm}:) modifies two sections of the input file.  First, tracers are added under the \texttt{Postprocess} section: 
+
+\lstinputlisting[language=prmfile]{cookbooks/composition-active-tracers/tracers.part.prm.out}
+
+In this case, the tracked tracer properties are \texttt{velocity} and \texttt{initial composition}.  In order to use the tracer initial composition value to modify the flow through the material model, we now modify the \texttt{Composition} section:
+
+\lstinputlisting[language=prmfile]{cookbooks/composition-active-tracers/composition.part.prm.out}
+ 
+In comparison to \url{cookbooks/composition-active.prm}, the \texttt{Composition} section now contains variables that specify the \texttt{Compositional field method} and \texttt{Mapped particle properties}.  The \texttt{Compositional field method} parameter may set to either \texttt{field} or \texttt{particles}, where \texttt{field} specifies that compositional fields are tracked by advected fields (default method).
```
Comment (gassmoeller) [inline doc/manual/manual.tex]: ... parameter may *be* set to either ...

File: doc/manual/manual.tex [hunk 6]
```diff
@@ -4745,6 +4745,26 @@ \subsubsection{Using tracer particles}
...
+
+In this case, we specify that the compositional fields are in fact tracked by particles.  In order for the material model to access information stored on the particles, however, we also need to specify that the initial composition of each particle is mapped to a new field created at each time step for use in the material model.  The name of this field will still correspond to those specified in \texttt{set Names of fields}.  
+
+The initial composition is mapped to a new field with the general phrase \texttt{field_name: initial field_name}, where \texttt{field_name} corresponds to \texttt{upper} or \texttt{lower}.  Notably, the order in which field names are defined in \texttt{set Names of fields} will determine the compositional field index number referenced in a material model.  However, it is recommended for consistency that mapped properties are listed in the same order as in \texttt{Set Mapped particle properties} .
```
Comment (gassmoeller) [inline doc/manual/manual.tex]: Maybe change the last sentence to : 
Therefore, it is recommended for consistency that mapped properties are listed in the same order as in \texttt{Set Names of fields}. ?

**Comment by gassmoeller** [comment]
Looks good, thanks for the nice work :+1:. Except for my two suggestions the text makes clear how to use the active tracers. Do you want to add a screenshot of the resulting model? Or is there no point in that, because it essentially looks like the active composition cookbook?

**Comment by naliboff** [comment]
Hi Rene - thanks for the review and glad everything looks good.  I incorporated your suggestions and a few other small fixes.  I originally had a screen shot of the resulting model, but removed it as it was indeed redundant with the current composition cookbook.  Will add that in later if folks find it helpful.  Otherwise, I think this should be good to go?

**Comment by gassmoeller** [comment]
Ok great. I will wait for the tester (should find nothing since its a cookbook), and then merge.

**Pull request merged**